crates/api/src/transport.rs

```diff
@@ -0,0 +1,350 @@
+//! Kitsune2 space related types.
+
+use crate::{protocol::*, *};
+use std::collections::HashMap;
+use std::sync::{Arc, Mutex};
+
+/// This is the low-level backend transport handler.
+/// Construct this ([TxImpHnd::new]) with a high-level [DynTxHandler],
+/// then call [TxImpHnd::gen_transport] to return the high-level handler
+/// from the [TransportFactory]. Then call the member functions at
+/// the appropriate times.
+pub struct TxImpHnd {
+    handler: DynTxHandler,
+    space_map: Arc<Mutex<HashMap<SpaceId, DynTxSpaceHandler>>>,
+    mod_map: Arc<Mutex<HashMap<(SpaceId, String), DynTxModuleHandler>>>,
+}
+
+impl TxImpHnd {
+    /// When constructing a [Transport] from a [TransportFactory],
+    /// you need a [TxImpHnd] for calling transport events.
+    /// Pass the handler into here to construct one.
+    pub fn new(handler: DynTxHandler) -> Arc<Self> {
+        Arc::new(Self {
+            handler,
+            space_map: Arc::new(Mutex::new(HashMap::new())),
+            mod_map: Arc::new(Mutex::new(HashMap::new())),
+        })
+    }
+
+    /// When constructing a [Transport] from a [TransportFactory],
+    /// This function does the actually wrapping of your implemementation
+    /// to produce the [Transport] struct.
+    pub fn gen_transport(&self, imp: DynTxImp) -> Transport {
+        Transport {
+            imp,
+            space_map: self.space_map.clone(),
+            mod_map: self.mod_map.clone(),
+        }
+    }
+
+    /// Call this when you receive or bind a new address at which
+    /// this local node can be reached by peers
+    pub fn new_listening_address(&self, this_url: Url) {
+        self.handler.new_listening_address(this_url);
+    }
+
+    /// Call this when you establish an outgoing connection and
+    /// when you establish an incoming connection. If this call
+    /// returns an error, the connection should be closed immediately.
+    /// On success, this function returns bytes that should be
+    /// sent as a preflight message for additional connection validation.
+    /// (The preflight data should be sent even if it is zero length).
+    pub fn peer_connect(&self, peer: Url) -> K2Result<bytes::Bytes> {
+        for h in self.mod_map.lock().unwrap().values() {
```

Comment:
Yeah, exactly. Gossip may want to know a peer is around so it can gossip with it. And some rate-limiting module may want to be able to kill incoming connections from bad peers.